src/eos-shard-shard-file-impl-v2.c

```diff
@@ -0,0 +1,259 @@
+    pread (self->fd, &blob, sizeof (blob), entry.blob_start);
+
+    char entry_name[EOS_SHARD_V2_BLOB_MAX_NAME_SIZE + 1] = {};
+    lookup_string_constant (impl, entry_name, sizeof (entry_name), blob.name_offs);
+    if (strncmp (entry_name, name, sizeof (entry_name)) == 0) {
+      *blob_out = blob;
+      return TRUE;
+    }
+  }
+
+  return FALSE;
+}
+
+static EosShardBlob *
+read_blob (EosShardShardFileImpl *impl, struct eos_shard_v2_record *srecord, const char *name)
+{
+  struct eos_shard_v2_blob blob;
+  if (!find_blob (impl, srecord, name, &blob))
+    return NULL;
+
+  return blob_new (impl, &blob);
+}
+
+static inline gboolean
+record_is_tombstone (struct eos_shard_v2_record *record)
+{
+  /* Tombstones indicate that records have been deleted.
+   * They're indicated with by a special flag in the record. */
+  return (record->flags & EOS_SHARD_V2_RECORD_FLAG_TOMBSTONE) != 0;
+}
+
+static EosShardRecord *
+record_new (EosShardShardFileImpl *impl, struct eos_shard_v2_record *srecord)
+{
+  EosShardShardFileImplV2 *self = EOS_SHARD_SHARD_FILE_IMPL_V2 (impl);
+
+  if (record_is_tombstone (srecord))
+    return NULL;
+
+  EosShardRecord *record = _eos_shard_record_new ();
+  record->shard_file = g_object_ref (self->shard_file);
+  record->raw_name = srecord->raw_name;
+  record->private_data = srecord;
+  record->metadata = read_blob (impl, srecord, EOS_SHARD_V2_BLOB_METADATA);
+  record->data = read_blob (impl, srecord, EOS_SHARD_V2_BLOB_DATA);
+  return record;
+}
+
+static int
+find_record_by_raw_name_compar (const void *a, const void *b)
+{
+  const struct eos_shard_v2_record *rec_a = a;
+  const struct eos_shard_v2_record *rec_b = b;
+  return memcmp (rec_a->raw_name, rec_b->raw_name, EOS_SHARD_RAW_NAME_SIZE);
+}
+
+static EosShardRecord *
+find_record_by_raw_name (EosShardShardFileImpl *impl, uint8_t *raw_name)
+{
+  EosShardShardFileImplV2 *self = EOS_SHARD_SHARD_FILE_IMPL_V2 (impl);
+  struct eos_shard_v2_record key, *res;
+
+  memcpy (key.raw_name, raw_name, EOS_SHARD_RAW_NAME_SIZE);
+
+  res = bsearch (&key, self->records, self->hdr.records_length, sizeof (*self->records),
+                 find_record_by_raw_name_compar);
+
+  if (res == NULL)
+    return NULL;
+  if (record_is_tombstone (res))
+    return NULL;
+
+  return record_new (impl, res);
+}
+
+static GSList *
+list_records (EosShardShardFileImpl *impl)
+{
+  EosShardShardFileImplV2 *self = EOS_SHARD_SHARD_FILE_IMPL_V2 (impl);
+  GSList *l = NULL;
+
+  int i = self->hdr.records_length;
+  while (i-- > 0) {
+    struct eos_shard_v2_record *srecord = &self->records[i];
+    if (record_is_tombstone (srecord))
+      continue;
+
+    EosShardRecord *record = record_new (impl, srecord);
+    l = g_slist_prepend (l, record);
+  }
+
+  return l;
+}
+
+static EosShardBlob *
+lookup_blob (EosShardShardFileImpl *impl, EosShardRecord *record, const char *name)
+{
+  struct eos_shard_v2_record *srecord = record->private_data;
+
+  /* Don't allow looking up internal record names. */
```

Comment:
Eh, I was iffy about this. I figured in case of a V3, I wouldn't want people attached to `$metadata`, but for something like `kermit`, it's already a bizarre special case.